File: 875. Koko Eating Bananas/code.java
class Solution {
    public int minEatingSpeed(int[] piles, int H) {
        
        int start = 1, end = 1_000_000_000;
        int max = end;
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            int count = calc(piles, mid);
            if (count > H) start = mid;
            else {
                if (mid + 1 < max && calc(piles, mid + 1) <= H) {
                    end = mid;
                    continue;
                }
                return start;
            }
        }
        
        if (calc(piles, start) <= H) return start;
        return end;
    }
    
    private int calc(int[] piles, double k) {
        int count = 0;
        for (int i : piles) {
            count += Math.ceil(i / k);
        }
        return count;
    }
}

File: 2218. Maximum Value of K Coins From Piles/code.cpp
class Solution {
public:
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
        int dp[2001] = {}, N = piles.size();
        for (int i = 0; i < N; ++i) {
            for (int j = k; j >= 1; --j) {
                for (int t = 1, sum = 0; t <= j && t <= piles[i].size(); ++t) {
                    sum += piles[i][t - 1];
                    dp[j] = max(dp[j], dp[j - t] + sum);
                }
            }
        }
        return dp[k];
    }
};
